addons/crm/models/crm_team.py

```diff
@@ -582,6 +560,82 @@ def _allocate_leads_deduplicate(self, leads, duplicates_cache=None):
+            * Assign the lead using round robin
+                * Find the first member with a compatible domain
+                * Assign the lead
+                * Move the member at the end of the list if quota is not reached
+                * Remove it otherwise
+                * Move to the next lead
+
+        :param bool force_quota: see ``CrmTeam._action_assign_leads()``;
+
+        :return members_data: dict() with each member assignment result:
+          membership: {
+            'assigned': set of lead IDs directly assigned to the member;
+          }, ...
+
+        """
+        auto_commit = not getattr(threading.current_thread(), 'testing', False)
+        result_data = {}
+        commit_bundle_size = int(self.env['ir.config_parameter'].sudo().get_param('crm.assignment.commit.bundle', 100))
+        teams_with_members = self.filtered(lambda team: team.crm_team_member_ids)
+        quota_per_member = {member: member._get_assignment_quota(force_quota=force_quota) for member in self.crm_team_member_ids}
+        counter = 0
+        leads_per_team = dict(self.env['crm.lead']._read_group(
+            [('user_id', '=', False), ('date_open', '=', False), ('team_id', 'in', teams_with_members.ids)],
+            ['team_id'],
+            ['id:recordset'],
+        ))
+        for team, leads_to_assign in leads_per_team.items():
+            members_to_assign = list(team.crm_team_member_ids.filtered(lambda member:
+                not member.assignment_optout and quota_per_member.get(member, 0) > 0
+            ).sorted(key=lambda member: quota_per_member.get(member, 0), reverse=True))
+            if not members_to_assign:
+                continue
+            result_data.update({
+                member: {"assigned": self.env["crm.lead"], "quota": quota_per_member[member]}
+                for member in members_to_assign
+            })
+            for lead in leads_to_assign.sorted(lambda lead: (-lead.probability, id)):
+                counter += 1
+                member_found = next((
+                    member for member in members_to_assign
+                    if lead in leads_to_assign.filtered_domain(literal_eval(member.assignment_domain or '[]'))), False)
+                if not member_found:
+                    continue
+                lead.with_context(mail_auto_subscribe_no_notify=True).convert_opportunity(
+                    lead.partner_id,
+                    user_ids=member_found.user_id.ids
+                )
+                result_data[member_found]['assigned'] += lead
+                members_to_assign.remove(member_found)
+                quota_per_member[member_found] -= 1
+                if quota_per_member[member_found] > 0:
+                    # If the member should receive more lead, send him back at the end of the list
+                    members_to_assign.append(member_found)
+
+                if auto_commit and counter % commit_bundle_size == 0:
+                    self._cr.commit()
```

Comment:
Side note: shouldn't we use self.env.cr, to avoid using the internal shortcuts ?